CVLByExamples/ConstantProductPoolExample/certora/spec/ConstantProductPool.spec

```diff
@@ -0,0 +1,264 @@
+
+/***
+This example explains many features of Certora Verification Language. 
+See https://docs.certora.com for a complete guide.
+***/
+
+// reference from the spec to additional contracts used in the verification 
+using DummyERC20A as _token0;
+using DummyERC20B as _token1;
+
+
+/*
+    Declaration of methods that are used in the rules. envfree indicate that
+    the method is not dependent on the environment (msg.value, msg.sender).
+    Methods that are not declared here are assumed to be dependent on env.
+*/
+
+methods{
+    function token0() external returns (address) envfree;
+    function token1() external returns (address) envfree;
+    function allowance(address,address) external returns (uint256) envfree;
+    function balanceOf(address) external returns (uint256)  envfree; 
+    function totalSupply() external returns (uint256)  envfree;
+    function getReserve0() external returns (uint256) envfree;
+    function getReserve1() external returns (uint256) envfree;
+    function swap(address tokenIn, address recipient) external returns (uint256) envfree;
+    
+    //calls to external contracts  
+    function _token0.balanceOf(address account) external returns (uint256) envfree;
+    function _token1.balanceOf(address account) external returns (uint256) envfree;
+    function _token0.transfer(address, uint) external;
+    function _token1.transfer(address, uint) external;
+
+    //external calls to be resolved by dispatcher - taking into account all available implementations 
+    function _.transferFrom(address sender, address recipient, uint256 amount) external => DISPATCHER(true);
+    function _.balanceOf(address) external => DISPATCHER(true);
+    
+}
+
+// a cvl function for precondition assumptions 
+function setup(env e){
+    address zero_address = 0;
+    uint256 MINIMUM_LIQUIDITY = 1000;
+    require totalSupply() == 0 || balanceOf(zero_address) == MINIMUM_LIQUIDITY;
+    require balanceOf(zero_address) + balanceOf(e.msg.sender) <= to_mathint(totalSupply());
```

Comment:
These two assumptions are very dangerous and should be loudly justified, or replaced with `requireInvariant`